### ISSUE TITLE
New rule: use-callback-hook

### ISSUE THREAD

**Issue by hornta** [yannickcr/eslint-plugin-react#2278]
This rule will report usages of independent standalone functions inside functional components and it will prevent unnecessary function redeclarations. By standalone function I mean functions that does not depend on either the state, props or context.

Consider this example when `handleClick` will be redeclared during every render, even though it is not needed.
```javascript
// invalid
function Foo() {
  const handleClick = (e) => { ... }
  return <button onClick={handleClick}>Button</button>
}
```

```javascript
// valid
function Foo() {
  const handleClick = useCallback((e) => { ... }, [ ... ]);
  return <button onClick={handleClick}>Button</button>
}
```

**Comment by ljharb** [comment]
When the function is being passed to a DOM component, it doesn't actually matter if it's redeclared every render.

**Comment by hornta** [comment]
Why doesn't it matter? What if it wasn't passed to a DOM element, but instead used in some other way inside the component? I think it's for the better to force memoization in this case.

**Comment by ljharb** [comment]
If it's passed to a DOM element, there's no downside and no performance hit. The only hazard is if it's passed to a custom component that's either a PureComponent, or has a `shouldComponentUpdate`, because it would cause unnecessary rerenders.

**Comment by fregadev** [comment]
@ljharb could that be a part of the rule, then? I find it that such a rule could be quite useful. I don't understand the reasoning behind pushing the idea aside instead of building on top of the problem it originally tries to solve.

**Comment by ljharb** [comment]
I think at this point, any hook-related rules should go in eslint-plugin-react-hooks, not in this package. Please feel free to file an issue there requesting it.
